Improving charm's documentation (#57)
canonical/loki-k8s-operator: CONTRIBUTING.md

## Overview

This documents explains the processes and practices recommended for
contributing enhancements or bug fixing to the Loki Charmed Operator.

## Setup

A typical development setup for charms created with the [Charmed Operator Framework](https://juju.is/docs/sdk) contains:

- [Charmcraft](https://github.com/canonical/charmcraft) - Developer tooling for creating, building and publishing Charmed Operators
- [Juju](https://juju.is/) - a Charmed Operator Lifecycle Manager (OLM), used for deploying and managing operators
- [Multipass](https://multipass.run/) - a lightweight Ubuntu virtual machine manager (optional)
- [MicroK8s](https://microk8s.io/) - a low-ops Kubernetes distribution we’ll use for testing our Charmed Operator (optional if developing a Charmed Operator for Kubernetes)

> Please [follow this guide](https://juju.is/docs/sdk/dev-setup) which will walk through the installation of these tools to get you started with charm development.

Author: We can do this, but in my view this is only a summary to guide the developer as to what to expect when they open that link.

What do you think @simskij ?

> Please [follow this guide](https://juju.is/docs/sdk/dev-setup) which will walk through the installation of these tools to get you started with charm development.

Reviewer: I'm torn. On one hand, I think it is indeed helpful to have this brief intro present in the charm context. On the other hand, it is another point where the instructions might start to diverge.

Let's leave it as-is for now and bring it up with the rest of the team on Tuesday's café.

  history.
- We use [`tox`](https://tox.wiki/en/latest/#) to manage all virtualenvs for the development lifecycle.

### Testing

All default tests can be executed by running `tox` without arguments.

You can also manually run specific test environment:

```bash
tox -e lint         # check your code complies to linting rules
tox -e static       # run static analysis
tox -e unit         # run unit tests
tox -e integration  # run integration tests
tox -e fmt          # update your code according to linting rules
```

Unit tests are written with the Operator Framework [test harness](https://ops.readthedocs.io/en/latest/#module-ops.testing).

### Build

In order to pack the charm locally so it could be deployed from a local path we use 
[charmcraft](https://juju.is/docs/sdk/setting-up-charmcraft).

From the charm's root folder:

```bash
$ charmcraft pack
Packing charm 'loki-k8s_ubuntu-20.04-amd64.charm'...
Created 'loki-k8s_ubuntu-20.04-amd64.charm'.
```

## Code Overview

### Charm code
This charm is represented by the [`LokiOperatorCharm`](src/charm.py) class, which
responds to the following events:

- `install`: Here we patch k8s service.

> - `config_change`: Here we configure the charm.

Reviewer 2: If you want to document the hook, do better. What do you configure? Loki process? Which parts? Under which conditions (e.g., established relations).

```bash
tox -e lint         # check your code complies to linting rules
tox -e static       # run static analysis
tox -e unit         # run unit tests
tox -e integration  # run integration tests
tox -e fmt          # update your code according to linting rules
```

Unit tests are written with the Operator Framework [test harness](https://ops.readthedocs.io/en/latest/#module-ops.testing).

### Build

In order to pack the charm locally so it could be deployed from a local path we use 
[charmcraft](https://juju.is/docs/sdk/setting-up-charmcraft).

From the charm's root folder:

```bash
$ charmcraft pack
Packing charm 'loki-k8s_ubuntu-20.04-amd64.charm'...
Created 'loki-k8s_ubuntu-20.04-amd64.charm'.
```

## Code Overview

### Charm code
This charm is represented by the [`LokiOperatorCharm`](src/charm.py) class, which
responds to the following events:

#### Juju Events

- `install`: Makes sure the k8s service is using the correct ports.
- `config_change`: Configures the charm.
- `loki_pebble_ready`: Sets up the pebble layer and starts the service
- `upgrade_charm`: Patches the ports of the k8s service (just as `install`) and configures the charm (just as `config_change`).

> - `alertmanager_consumer.on.cluster_changed`: Event provided by the `AlertmanagerConsumer`. Further configures the charm to integrate with Alertmanager.

Reviewer 3: ~~the~~ `AlertmanagerConsumer`

  The primary author of this charm is available on the Mattermost channel as

>   `@jose-masson`.

Author: Please @simskij let us know your thoughts about this

>   `@jose-masson`.

Reviewer: I think this is sufficient. If anything, you could add a sentence about the rest of the canonical observability team being available and open to help in the ~observability channel on mattermost as well.

- It is strongly recommended that prior to engaging in any enhancements

>   to this charm you familiarise your self with [Juju](https://juju.is).

Author: What do you think if we link to this section? https://juju.is/#what-is-juju

>   to this charm you familiarise your self with [Juju](https://juju.is).

Reviewer: Guess it could be phrased in multiple ways. One way to phrase it that I think is a bit more inviting is

> To make sure you get a smooth, worthwhile experience contributing to this charm, we recommend that you first familiarize yourself with Juju. We also recommend that you join the CharmHub Mattermost to get early feedback on your ideas.

>   to this charm you familiarise your self with [Juju](https://juju.is).

Reviewer 3: Throwing another alternative to the mix:

> To author PRs you should know [what is jujuj](https://juju.is/#what-is-juju) and [how operators are written](https://juju.is/docs/sdk). The best way to get a head start is to join the conversation on [our mattermost channel](https://chat.charmhub.io/charmhub/channels/observability)!

  this charm. Please help us out in having easier reviews by rebasing
  onto the `main` branch, avoid merge commits and enjoy a linear Git
  history.

> - To handle our virtualenvs we use [`tox`](https://tox.wiki/en/latest/#), so there is no need
>   to deal with them manually. Tox will create, update, activate and deactivate virtualenvs for us.

Reviewer 3: ```suggestion
- We use [`tox`](https://tox.wiki/en/latest/#) to manage all virtualenvs for the development lifecycle.
```

### Testing

> All tests can be executed by running `tox` without arguments.

Reviewer 3: ```suggestion
All default tests can be executed by running `tox` without arguments.
```

> Besides you can run individual test environments or tasks:

Reviewer 3: ```suggestion
You can also manually run specific test environment:
```

```bash
tox -e lint         # check your code complies to linting rules
tox -e static       # run static analysis
tox -e unit         # run unit tests
tox -e integration  # run integration tests
tox -e fmt          # update your code according to linting rules
```

> Unit tests are implemented using the Operator Framework test [harness](https://ops.readthedocs.io/en/latest/#module-ops.testing).

Reviewer 3: ```suggestion
Unit tests are written with the Operator Framework [test harness](https://ops.readthedocs.io/en/latest/#module-ops.testing).
```

### Build

> In order to build the charm so it can be deployed in [MicroK8s](https://microk8s.io/) using [Juju](https://juju.is/), we use [charmcraft](https://juju.is/docs/sdk/setting-up-charmcraft).
> So in the charm repository you have to run:

Reviewer 3: ```suggestion
In order to pack the charm locally so it could be deployed from a local path we use 
[charmcraft](https://juju.is/docs/sdk/setting-up-charmcraft).

From the charm's root folder:
```

```bash
$ charmcraft pack
Packing charm 'loki-k8s_ubuntu-20.04-amd64.charm'...
Created 'loki-k8s_ubuntu-20.04-amd64.charm'.
```

## Code Overview

> The core implementation of this charm is represented by the [`LokiOperatorCharm`](src/charm.py) class, which
> responds to the following events:

Reviewer 3: ```suggestion
### Charm code
This charm is represented by the [`LokiOperatorCharm`](src/charm.py) class, which
responds to the following events:
```

- `install`: Here we patch k8s service.
- `config_change`: Here we configure the charm.
- `loki_pebble_ready`: Here we set up pebble layer and start the service
- `upgrade_charm`: Here we patch k8s service and configure the charm.

> - `alertmanager_consumer.on.cluster_changed`: This event is provided by the AlertmanagerConsumer object. Here we configure the charm.

Reviewer 3: ```suggestion
- `alertmanager_consumer.on.cluster_changed`: This event is provided by the AlertmanagerConsumer object. Here we configure the charm.

### Library code
TODO
```

Author: The Loki libraries' code is explained in INTEGRATING.md file

Reviewer: The suffix "object" adds no additional value. See suggestion above.
